components/omega/src/ocn/Tracers.cpp

```diff
@@ -0,0 +1,524 @@
+// Define tracers
+//---------------------------------------------------------------------------
+I4 Tracers::define(const std::string &Name, const std::string &Description,
+                   const std::string &Units, const std::string &StdName,
+                   const Real ValidMin, const Real ValidMax,
+                   const Real FillValue) {
+
+   // Do nothing if this tracer is not selected
+   if (TracerIndexes.find(Name) == TracerIndexes.end()) {
+      return 0;
+   }
+
+   auto TracerIndex = TracerIndexes[Name];
+
+   // Return error if tracer already exists
+   if (TracerNames.find(TracerIndex) != TracerNames.end()) {
+      LOG_ERROR("Tracers: Tracer '{}' already exists", Name);
+      return -1;
+   }
+
+   // set tracer index to name mapping
+   TracerNames[TracerIndex] = Name;
+
+   // create a tracer field
+   std::string TracerFieldName = Name;
+   auto TracerField = Field::create(TracerFieldName, Description, Units,
+                                    StdName, ValidMin, ValidMax, FillValue,
+                                    TracerDimNames.size(), TracerDimNames);
+   if (!TracerField) {
+      LOG_ERROR("Tracers: Tracer field '{}' is not created", TracerFieldName);
+      return -2;
+   }
+
+   return 0;
+}
+
+//---------------------------------------------------------------------------
+// Deallocate tracer arrays
+//---------------------------------------------------------------------------
+I4 Tracers::clear() {
+
+   LOG_INFO("Tracers::clear() called");
+
+   // Deallocate memory for tracer arrays
+   TracerArrays.clear();
+   TracerArraysH.clear();
+
+   TracerGroups.clear();
+   TracerIndexes.clear();
+   TracerNames.clear();
+
+   NumTracers  = 0;
+   NCellsOwned = 0;
+   NCellsAll   = 0;
+   NCellsSize  = 0;
+   NTimeLevels = 0;
+   NVertLevels = 0;
+
+   return 0;
+}
+
+//---------------------------------------------------------------------------
+// Query tracers
+//---------------------------------------------------------------------------
+
+I4 Tracers::getNumTracers() { return NumTracers; }
+
+I4 Tracers::getIndex(I4 &TracerIndex, const std::string &TracerName) {
+   // Check if tracer exists
+   if (TracerIndexes.find(TracerName) != TracerIndexes.end()) {
+      TracerIndex = TracerIndexes[TracerName];
+      return 0; // Success
+   }
+
+   LOG_ERROR("Tracers: Tracer index for '{}' is not found.", TracerName);
+   return -1; // Tracer not found
+}
+
+I4 Tracers::getName(std::string &TracerName, const I4 TracerIndex) {
+   if (TracerNames.find(TracerIndex) != TracerNames.end()) {
+      TracerName = TracerNames[TracerIndex];
+      return 0; // Success
+   }
+
+   LOG_ERROR("Tracers: Tracer name for index '{}' is not found.", TracerIndex);
+   return -1; // Tracer index not found
+}
+
+//---------------------------------------------------------------------------
+// get Tracer arrays
+// TimeLevel == [0:current, -1:previous, -2:two times ago, ...]
```

Comment:
Thanks for bringing this issue up. I think it's worth some discussion because the difference in indexing conventions between Tracers and State stems from how the time level index is managed.

Both State and Tracers have a std::vector (let's call it ARR) as a private member in their respective classes to maintain multiple time levels.

In State, NextLevel is always set to the highest index of ARR, and CurLevel is NextLevel - 1. Since the ARR indices for the time levels are fixed, to increment the time level in State, the elements of ARR are swapped. I guess this approach may be better aligned with the established convention.

In Tracers, a private integer member (let's call it IDX) keeps track of the array (ARR) index for the current time level. To increment the time level, IDX is simply increased by 1. If IDX exceeds the size of ARR, it is reset to zero. I chose this approach to maximize performance because adding 1 is computationally cheap.

Reply:
> Both State and Tracers have a std::vector (let's call it ARR) as a private member in their respective classes to maintain multiple time levels.

Small correction, state stores its variables as a public `Kokkos::Array<Array2DReal, MaxTimeLevels>`.

Reply:
I explained why I had suggested using `Kokkos::Array` for the State in the response to @sbrus89 comment below and why I don't think it is necessary. As for the index convention, it is a separate issue, and I don't have a strong preference. One thing that would be nice is to have a common function for checking that a time index is valid.

Reply:
@grnydawn and @mwarusz, after looking at this a little more, I think we should change State to follow the time index convention here (as well as using the std::vector approach).